dedalus/core/solvers.py

```diff
@@ -144,99 +138,112 @@ class EigenvalueSolver(SolverBase):
     def __init__(self, problem, **kw):
         logger.debug('Beginning EVP instantiation')
         super().__init__(problem, **kw)
-        # Build subsystems and subproblem without building matrices
-        self.subsystems = subsystems.build_subsystems(self)
-        self.subproblems = subsystems.build_subproblems(self, self.subsystems)
-        self.subproblems_by_group = {sp.group: sp for sp in self.subproblems}
-        self.state = problem.variables
         logger.debug('Finished EVP instantiation')
 
-    def solve_dense(self, subproblem, rebuild_coeffs=False, **kw):
+    def print_subproblem_ranks(self, subproblems=None, target=0):
+        """Print rank of each subproblem LHS."""
+        if subproblems is None:
+            subproblems = self.subproblems
+        # Check matrix rank
+        for i, sp in enumerate(subproblems):
+            if not hasattr(sp, 'L_min'):
+                continue
+            L = (sp.L_min @ sp.pre_right).A
+            M = (sp.M_min @ sp.pre_right).A
+            A = L + target * M
+            print(f"MPI rank: {self.dist.comm.rank}, subproblem: {i}, group: {sp.group}, matrix rank: {np.linalg.matrix_rank(A)}/{A.shape[0]}, cond: {np.linalg.cond(A):.1e}")
+
+    def _build_modified_left_eigenvectors(self):
+        sp = self.eigenvalue_subproblem
+        return - (self.left_eigenvectors.T.conj() * sp.M_min).T.conj()
+
+    def _normalize_left_eigenvectors(self):
+        modified_left_eigenvectors = self._build_modified_left_eigenvectors()
+        norms = np.diag(modified_left_eigenvectors.T.conj() @ self.eigenvectors)
+        self.left_eigenvectors /= norms
+
+    def solve_dense(self, subproblem, rebuild_matrices=False, left=False, normalize_left=True, **kw):
         """
-        Solve EVP for selected pencil.
+        Perform dense eigenvector search for selected subproblem.
+        This routine finds all eigenvectors but is computationally expensive.
 
         Parameters
         ----------
-        subproblem : subproblem object
-            Subproblem for which to solve the EVP
-        #rebuild_coeffs : bool, optional
-        #    Flag to rebuild cached coefficient matrices (default: False)
-
-        Other keyword options passed to scipy.linalg.eig.
-
+        subproblem : Subproblem object
+            Subproblem for which to solve the EVP.
+        rebuild_matrices : bool, optional
+            Rebuild LHS matrices if coefficients have changed (default: False).
+        left : bool, optional
+            Solve for the left eigenvectors of the system in addition to the
+            right eigenvectors. The left eigenvectors are the right eigenvectors
+            of the conjugate-transposed problem. Follows same definition described
+            in scipy.linalg.eig documentation (default: False).
```

Comment:
Should we mention we're multiplying by -M? I don't think that is part of the scipy.linalg.eig documentation.